text/0006-shipyard-next-generation.md

```diff
@@ -0,0 +1,317 @@
+    * (*) Domain event: *An event that occurred in the business process, written in past tense; [see](https://en.wikipedia.org/wiki/Event_storming)*. This is fired by a human or tool to inform about a certain situation. For example, a `problem.open` event is fired by a monitoring tool when a service runs in a problem mode.
+
+1. Task events: For each task, a `[task].triggered` event is sent by the control plane. Those Keptn-services that have a subscription on this event, will react with a `[task].started` event, perform their functions, and finally confirm their execution with a `[task].finished`. E.g., for the *test* task, the following events occurre:
+    * `test.triggered`
+      * `test.started`
+      * `test.finished`
+
+1. Task properties: The additional task properties (e.g., task `test` specifies the property: `kind:functional`) is added to the `[task].triggered` event.
+
+1. Data flow: A task can add any data to the event payload (reserved space, see below). The Keptn's control plane takes care that the data is not lost and delivered to the next task. As more tasks are added to a sequence, as more data can be added to the event stream (= all events occurring in a task sequence).
+
+### Event selector
+
+A sequence has a subscription to certain events that trigger the sequence. However, an SRE not only wants to configure the event type, but also a selector on this event. This selector configures more precisely when the sequence should be triggered. For example, an SRE wants to configure that a rollback sequence should be triggered just when deployment or release failed. Therefore, this KEP proposes the approach of an event selector using  `matchLabels` and `matchExpressions`. 
+
+While the `matchLabels` allows configuring a selector on a `key` with exactly one expected `value`, the `matchExpressions` allows configuring a selector on a `key` with multiple value options. For example, the below rollback sequence has a subscription to two events. (1) For the first `hardening.deployment.finished` event a selector is configured based on the label `status:failed`. This means that the rollback sequence gets triggered when the event has the label `status:failed` attached. (2) For the second `hardening.release.finished` event a selector is configured for the key `status`, which can have either the value `warning` or `failed`.
+
+```yaml
+rollback:
+  triggers:
+  - hardening.deployment.finished:
+      selector:
+        matchLabels:
+          status: failed
+  - hardening.release.finished:
+      selector:
+        matchExpressions:
+          - {key: status, operator: In, values: [warning, failed]}
+  tasks:
+  - name: rollback
+```
```

Comment:
This has not made it into the spec yet. Also, did we implement this or do we still need an issue for it?

Reply:
I filed an issue for that: https://github.com/keptn/keptn/issues/3028